### PR TITLE
Alsa widget

### PR DESCRIPTION
**Pull request by Ploppz**
Listens for alsa events, then polls the volume and muted state from alsa.

Currently uses Linux' poll because it was the easiest.

Not tested extensively and only on my computer.

### CODE REVIEW

**Comment by valpackett** [comment]
Nice. But you should add pluggable backend support, i.e. `Volume::new(ALSA::new(), |vol| …)`, and the ALSA backend should be in a module that'd only be built on Linux. (And the Cargo feature for ALSA should also be Linux specific.)

**Comment by Ploppz** [comment]
Like this? Feel free to review critically, I would like to know how to do it better.

File: Cargo.toml [hunk 1]
```diff
@@ -22,8 +22,11 @@ pkg-config = { version = "0.3", optional = true }
 chrono = "0.2"
 nom = "2.0"
 systemstat = { version = "0", optional = true }
-xcb = { version = "0.7", features = ["xkb"], optional = true }
+xcb = { version = "0.8", features = ["xkb"], optional = true }
```
Comment (Ploppz) [inline Cargo.toml:24]: 0.7 doesn't compile on Nightly

File: Cargo.toml [hunk 2]
```diff
@@ -21,6 +22,11 @@ pkg-config = { version = "0.3", optional = true }
...
 serde = "0.8"
 serde_json = "0.8"
+alsa = { version = "0.1.10", optional = true }
+libc = { version = "0.2.29", optional = true }
+
+[patch.crates-io]
```
Comment (valpackett) [inline Cargo.toml]: hmm what's this?

File: src/widget/volume/mod.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+use std::sync::Arc;
+use widget::base::Widget;
+use format::data::Format;
+use std::sync::mpsc::Sender;
+
+#[cfg(unix)] pub mod alsa;
+#[cfg(unix)] pub use self::alsa::ALSA;
```
Comment (valpackett) [inline src/widget/volume/mod.rs]: `#[cfg(target_os = "linux")]`. It's the Advanced *Linux* Sound Architecture after all.

File: Cargo.toml [hunk 4]
```diff
@@ -11,8 +11,9 @@ repository = "https://github.com/myfreeweb/unixbar"
 build = "build.rs"
 
 [features]
-default = ["systemstat", "xkb"]
+default = ["systemstat", "xkb", "unix_alsa"]
 xkb = ["xcb", "pkg-config"]
+unix_alsa = ["alsa", "libc"]
```
Comment (valpackett) [inline Cargo.toml]: something like 

```
[target.'cfg(target_os = "linux")'.dependencies]
default = ["systemstat", "xkb", "unix_alsa"]
```

(and no alsa in the… default `default` :D)

**Review by valpackett — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 2 above)
Comment (Ploppz) [reply Cargo.toml]: again patch.crates-io snuck in unintentionally, I will remove it.

Comment (valpackett) [reply Cargo.toml]: But what *is* the `patch` thing? Never seen that before :)

File: Cargo.toml [hunk 5]
```diff
@@ -25,8 +24,7 @@ systemstat = { version = "0", optional = true }
 xcb = { version = "0.8", features = ["xkb"], optional = true }
 serde = "0.8"
 serde_json = "0.8"
-alsa = { version = "0.1.10", optional = true }
-libc = { version = "0.2.29", optional = true }
 
-[patch.crates-io]
-systemstat = { git = "https://github.com/myfreeweb/systemstat" }
+[target.'cfg(target_os = "linux")'.dependencies]
```
Comment (Ploppz) [inline Cargo.toml]: Removed the `unix_alsa` feature, this seems simpler.

Comment (valpackett) [reply Cargo.toml]: Yeah that's fine for now, though I'd imagine someone might prefer e.g. a pulseaudio-only build in the future.

**Review by valpackett — APPROVED** [review]

**Comment by valpackett** [comment]
okay thanks

File: Cargo.toml (hunk 2 above)
Comment (Ploppz) [reply Cargo.toml]: Afaik, it lets me temporarily change the source of some dependency. Specifically I needed linux cpu support from `systemstat` on github (perhaps there should be a new release?). I don't really remember why options that went into .cargo/config didn't work for me.

**Pull request merged**